### ISSUE TITLE
Automerge.diff performance calling clock_at

### ISSUE THREAD

**Issue by stephen** [automerge/automerge#652]
I was previously doing `Automerge.applyChanges(Automerge.init(), Automerge.getAllChanges(), { patchCallback })` to look at the patches from my document but then saw `Automerge.diff` from https://github.com/automerge/automerge/pull/612 which was very useful!

This took my patch-compute time down from 2.1s to 850ms, mostly due to not allocating and moving strings back and forth through the wasm/js barrier).

However, ~850ms still seems like a relatively long time (3ms/change) on my medium-sized document (~219kb, ~300changes on an m1 macbook air).

<img width="779" alt="Screenshot 2023-06-24 at 1 51 06 PM" src="https://github.com/automerge/automerge/assets/2320890/ea4b8034-b384-44d8-8adb-085bbe73fffb">

Looking at the stack trace, most of this time seems to be spent calling [`clock_at`](https://github.com/automerge/automerge/blob/main/rust/automerge/src/automerge.rs#L1460) (from the `get_path` call):

<img width="975" alt="Screenshot 2023-06-24 at 1 54 55 PM" src="https://github.com/automerge/automerge/assets/2320890/87d59a6d-3225-49ad-983d-123d18964215">

I wonder if the clocks here could be cached for the diff call? With my limited rust experience, I haven't investigated if this code is actually repeatedly calling `clock_at` for the same heads, but that's my current guess.

I did see https://github.com/automerge/automerge/commit/13a775ed9adc04c55067e3dc2eaa294fc862cb09 which looks like it does the opposite (compute the clock on demand instead of up front for loads).

If this seems like a reasonable line of thinking around the expected performance, I'd be happy to dig more.

**Comment by Geo25rey** [comment]
Related to #644

**Comment by alexjg** [comment]
The reason we can't pre-compute the clocks is because it torpedoes load and `receiveSyncMessage` times with large numbers of changes. However, you are right, `changeAt` currently suffers a large performance decrease vs `change` due to the need for `clock_at`. The plan here is to follow the road begun in `changeAt` a little further and build branches. Branches are collections of heads with long lived possibly divergent changes which you can use independently of each other. In order to do this efficiently we will have a `git checkout` style API which will build a set of indexes allowing fast edits. So where you currently use `changeAt` you would instead do something like

```typescript
const doc = ...
let branch = automerge.checkout(doc, <some heads>)
branch = automerge.change(branch, ...)
```

(Note that this is literally something I just made up and not the result of thinking hard about API design)

This work is next on our list once we've got Peritext with blocks implemented and out the door. I'm working on writing up some design documents for branches and related work which I'll publish here and in the slack.

**Comment by alexjg** [comment]
@Geo25rey I don't _think_ this is related to #644 but do you have reasons to think it is?

**Comment by stephen** [comment]
@alexjg Thanks for the context! I am not actually using `changeAt` but just `diff` like...
```js
const patches = Automerge.diff(doc, [], Automerge.getHeads(doc));
for (const p of patches) {
  // ...
}
```

but it sounds like with the upcoming checkout api the "currently checked out heads" metadata could be fitted with a index of hash -> clock that can speed up the `diff` call - does that sound right? something like:

```ts
const checkout = Automerge.checkout(doc, Automerge.getHeads(doc));
const patches = checkout.diff([]); // diff against empty heads / start of document
```

**Comment by alexjg** [comment]
Ah, I've misunderstood what you were aiming to do as my slightly sleep deprived brain thought you were referring to `changeAt` rather than `diff` :grimacing: 

What do you want the patches for exactly? The patches returned by `diff` won't necessarily reflect a particular set of changes, they're just the patches require to get you from one set of heads to another. In the case of starting with empty heads this would be the patches required to construct the state you get from `automerge.load`.

Regardless, it should be possible to make `diff` fast in the case of diffing against empty heads as there's no need to calculate clocks there. I'll take a look.

**Comment by stephen** [comment]
> What do you want the patches for exactly?

I have some code that follows patches to the automerge document and tails them into a sqlite database. On normal changes, I do:
```ts
Automerge.change(doc, { patchCallback: patchCallback }, (doc) => {
  doc.records["id-1"] = { value: 3, etc: "hello };
})
```

and the callback takes each record and puts it in the `records` table of the db ([some old prototype code if it helps](https://gist.github.com/stephen/d6603c058f0a6ea833afd5a415bb9859#file-tailer-ts-L113)).

Sometimes, the sqlite db schema changes and I need to throw away/rebuild the db from the beginning of the document. Previously I was doing:
```ts
Automerge.applyChanges(
  Automerge.init(),
  Automerge.getAllChanges(doc),
  {
    patchCallback: patchCallback,
  }
);
```
where I'm doing a wasteful thing (applying on the changes to a new doc) just to get the `patchCallback` side effect, but now that `diff` exists, I can do:

```ts
const patches = Automerge.diff(doc, [], Automerge.getHeads(doc));
patchCallback(patches)
```

**Comment by alexjg** [comment]
Ah okay so the value of getting all the patches instead of just using the state directly is in only having to maintain one codepath which puts stuff in the database?

**Comment by stephen** [comment]
yes - i didn't want to have a separate codepath for "rebuild from the current document" and "build from incremental change"

**Comment by alexjg** [comment]
Makes sense.

So, you _should_ be hitting the fast path (no clock generation) with `automerge.diff(doc, [], automerge.getHeads(doc)))`, the fact that you aren't is a bug. If you have an easy reproduction of this handy that would be very helpful, otherwise I'll have a bit of a dig.

**Comment by stephen** [comment]
here's a quick repro: https://gist.github.com/stephen/31d8fb82abfa0d8ae65f0cb5d094bb6e

it looks like `diff` time is quadratic with changes

**Comment by alexjg** [comment]
@stephen the PR above makes the `5000` change diff take 15ms on my machine.

**Comment by stephen** [comment]
@alexjg thanks for taking a look! I tried the branch out on my code and it looks around ~85% faster for my app 🎉 (~800ms down to ~120ms)

on the benchmark gist above i see the same result as you (was ~2600ms down to ~25ms).

my actual document has more complex stuff going on so I suppose it's not unreasonable that the improvement isn't 100x.. from the trace looks like it spends 2/3 of that time crossing js <> wasm land:
<img width="950" alt="Screenshot 2023-06-26 at 5 11 25 PM" src="https://github.com/automerge/automerge/assets/2320890/77a22c23-8f2e-4f59-aa0c-df8ad99a8833">

**Comment by alexjg** [comment]
Yeah I think we'll probably need to wait for improvements to the WASM/JS integration to fix that particular problem. I'm glad we were able to cut it down so much though.

The fix is published in `@automerge/automerge@2.1.0-alpha.7` BTW.
